src/compile.rs

```diff
@@ -320,6 +324,205 @@ pub fn validate_functions<'a, 'b>(
+                        &format!("Resources within modules must be declared elsewhere"),
+                    ),
+                )),
+            }
+        }
+
+        for m in &module.modules {
+            if !&modules_vec.iter().any(|&x| x.1.name == m.as_ref()) {
+                errors.append(CascadeErrors::from(
+                    ErrorItem::make_compile_or_internal_error(
+                        &format!("Module {} does not exist", m.as_ref()),
+                        Some(file),
+                        m.get_range(),
+                        &format!("Modules within modules must be declared elsewhere"),
+                    ),
+                ))
+            }
+        }
+    }
+    errors = errors.into_result_self()?;
+
+    // Make sure there are no cycles in the modules
+    for m in &modules_vec {
+        match find_module_cycles(m.1, &modules_vec, HashSet::new()) {
+            Ok(()) => (),
+            Err(e) => errors.append(e),
+        }
+        errors = errors.into_result_self()?;
+    }
+
+    // Create Validated Modules
+    let mut validated_modules: Vec<ValidatedModule> = Vec::new();
+    for m in &modules_vec {
+        if !&validated_modules.iter().any(|x| x.name == m.1.name) {
+            validated_modules =
+                create_validated_modules(m.1, validated_modules.clone(), &modules_vec, types);
+        }
+    }
+    errors.into_result(())
+}
+
+fn find_module_cycles<'a>(
+    module_to_check: &Module,
+    modules_vec: &Vec<(&SimpleFile<String, String>, &'a Module)>,
+    visited_modules: HashSet<&str>,
+) -> Result<(), CascadeErrors> {
+    let mut ret = CascadeErrors::new();
+    for m in &module_to_check.modules {
+        match modules_vec
+            .iter()
+            .find(|&module_info| module_info.1.name == m.as_ref())
+        {
+            Some(module_info) => {
+                if visited_modules.contains(m.as_ref()) || *m == module_to_check.name {
+                    // Cycle
+                    return Err(CascadeErrors::from(
+                        ErrorItem::make_compile_or_internal_error(
+                            "Cycle detected",
+                            Some(module_info.0),
+                            m.get_range(),
+                            "This module contains itself or has a descendent that contains it",
+                        ),
+                    ));
+                }
+                let inner_module = module_info.1;
+
+                let mut new_visited_modules = visited_modules.clone();
+                new_visited_modules.insert(module_to_check.name.as_ref());
+
+                match find_module_cycles(inner_module, modules_vec, new_visited_modules) {
+                    Ok(()) => (),
+                    Err(e) => ret.append(e),
+                }
+            }
+            None => (),
+        }
+    }
+    ret.into_result(())
+}
+
+fn create_validated_modules<'a>(
+    curr_module: &'a Module,
+    mut created_modules: Vec<ValidatedModule>,
```

Comment:
This whole function feels like a recursive function that's really trying to be an iterative function.  Each call to create_validated_modules() takes all the modules so far, and then returns a complete list.

I've been going back and forth about a cleaner approach, since I feel like this whole clone/recursion thing will be difficult to maintain.

What if we created the modules initially with just the domains and resources and left the child modules as an empty Vec.  Then do a second pass, where we can just straightforwardly look up the already built ValidatedModules and store references to them in the Vec?  I think that will be a lot clearer and cleaner.